File: src/vm/mod/iodev/probe.cpp
#include "probe.hpp"

namespace kcpu::iodev {

probe::probe(const std::unordered_map<regval_t, std::reference_wrapper<io_device>> &ports) : single_port_io_device(PORT_BASE), target_port(0), ports(ports) {
}

halfcycle_count_t probe::write(regval_t val) {
    target_port = val;
    return 0;
}

std::pair<regval_t, halfcycle_count_t> probe::read() {
    // This assertion is not formally neccesary, but is useful to detect bad port IO.
    // i.e. when we read from port 0 by accident, when we aren't using the probe function.
    vm_assert(target_port != 0);
    return std::pair(ports.find(target_port) == ports.end() ? 0 : 1, 0);
}

}

File: src/vm/io/dev/probe.rs
use super::super::types::*;
use crate::spec::types::hw::*;

const PORT_BASE: Word = 0x00;

pub struct Probe {
    target_port: Word,
    // RUSTFIX expose a way to access this data being stored in the manager, so that we can change these at runtime
    ports: Vec<Word>,
}

impl Probe {
    pub fn new(ports: Vec<Word>) -> Self {
        Probe {
            target_port: 0,
            ports,
        }
    }
}

impl SinglePortDevice for Probe {
    fn reserved_port(&self) -> Word {
        PORT_BASE
    }

    fn write(&mut self, val: Word) -> HalfcycleCount {
        self.target_port = val;
        0
    }

    fn read(&mut self) -> (HalfcycleCount, Word) {
        // This assertion is not formally neccesary, but is useful to detect bad port IO.
        // i.e. when we read from port 0 by accident, when we aren't using the probe function.
        assert_ne!(self.target_port, 0);
        (
            0,
            if self.ports.contains(&self.target_port) {
                1
            } else {
                0
            },
        )
    }

    fn process_halfcycle(&mut self, _: ClockedSignals) {}
}
